libcontainer/container_linux.go

```diff
@@ -1235,11 +1235,39 @@ func (c *linuxContainer) prepareCriuRestoreMounts(mounts []*configs.Mount) error
 	// Now go through all mounts and create the mountpoints
 	// if the mountpoints are not on a tmpfs, as CRIU will
 	// restore the complete tmpfs content from its checkpoint.
+	umounts := []string{}
+	defer func() {
+		for _, u := range umounts {
+			if e := unix.Unmount(u, unix.MNT_DETACH); e != nil {
+				if e != unix.EINVAL {
+					// Ignore EINVAL as it means 'EINVAL target is not a mount point.'
+					// It probably has already been unmounted.
+					logrus.Warnf("Error during cleanup unmounting of %q (%v)\n", u, e)
+				}
+			}
+		}
+	}()
 	for _, m := range mounts {
 		if !isPathInPrefixList(m.Destination, tmpfs) {
 			if err := c.makeCriuRestoreMountpoints(m); err != nil {
 				return err
 			}
+			// If the mount point is a bind mount, we need to mount
+			// it now so that runc can create the necessary mount
+			// points for mounts in bind mounts.
+			// This also happens during initial container creation.
+			// Without this CRIU restore will fail
+			// See: https://github.com/opencontainers/runc/issues/2748
+			// It is also not necessary to order the mount points
+			// because during initial container creation mounts are
+			// set up in the order they are configured.
+			if m.Device == "bind" {
+				if err := unix.Mount(m.Source, m.Destination, "", unix.MS_BIND|unix.MS_REC, ""); err != nil {
```

Comment:
We _might_ need to resolve m.Destination inside a container root (see e.g. https://github.com/opencontainers/runc/pull/2715/commits/637f82d6b8b45a99519b863f8cf7461760da7a97 or `prepareBindMount` at https://github.com/opencontainers/runc/blob/94ae7afa36cc3b8f551e0bc67cf83e5efdf2555f/libcontainer/rootfs_linux.go#L198-L212)

To check, you need to add a test case where `m.Destination` is a symlink (something like `/mnt/test2 -> ../test1` or `/test2 -> /test1`).

Reply:
I was thinking about the exactly the same thing and initially had the code copied from `prepareBindMount()`, but as  `prepareBindMount()` is updating `m.Destination` with the changed destination path (https://github.com/opencontainers/runc/blob/master/libcontainer/rootfs_linux.go#L217) there is no need to run it a second time here.